### PR TITLE
Extend ABCI `max_block_size` parameter to give extended control to the app

### LINKED ISSUE #980: ABCI++: application to control how many txs it receives in PrepareProposal

### PR DESCRIPTION
**Pull request by sergio-mena**
Closes #980

This PR is the result of internal discussions (last year, but also last week), and interactions with the Vega team in order to address their use case.

The gist of the agreed solution is to interpret value -1 for `ConsensusParams.Block.MaxBytes` as the Application wanting to enforce a size limit at its level, via `PrepareProposal` and `ProcessProposal`. Additionally, it instructs CometBFT to send all transactions in the mempool upon calling `PrepareProposal` so that the application can apply its own transaction ordering policies to *all* transactions known to the node.

This solution is also the most _surgical_ among all candidate solutions in that the least amount of logic/spec needed to be added/removed/modified.

So, at CometBFT level, a value of -1 for `ConsensusParams.Block.MaxBytes` implies:

* The limit enforced at CometBFT level is the generic `types.MaxBlockSizeBytes` (hard-coded to 100MB)
* CometBFT will reap all transactions from the mempool when creating a proposed block (just before calling `PrepareProposal`)

---

#### PR checklist

- [x] Tests written/updated
- [x] Changelog entry added in `.changelog` (we use [unclog](https://github.com/informalsystems/unclog) to manage our changelog)
- [x] Updated relevant documentation (`docs/` or `spec/`) and code comments

### CODE REVIEW

**Review by cason — APPROVED** [review]
Legit.

File: types/params.go [hunk 1]
```diff
@@ -143,8 +143,11 @@ func IsValidPubkeyType(params ValidatorParams, pubkeyType string) bool {
 // Validate validates the ConsensusParams to ensure all values are within their
 // allowed limits, and returns an error if they are not.
 func (params ConsensusParams) ValidateBasic() error {
-	if params.Block.MaxBytes <= 0 {
-		return fmt.Errorf("block.MaxBytes must be greater than 0. Got %d",
+	if params.Block.MaxBytes == 0 {
+		return fmt.Errorf("block.MaxBytes cannot be 0")
+	}
+	if params.Block.MaxBytes < -1 {
+		return fmt.Errorf("block.MaxBytes must be greater than -1. Got %d",
```
Comment (lasarojc) [inline types/params.go:150]:
```suggestion
		return fmt.Errorf("block.MaxBytes must be -1 or greater than 0. Got %d",

```

**Review by lasarojc — APPROVED** [review]
Some nitpicking, but LGTM.

File: .changelog/unreleased/improvements/980-max-size-more-control.md [hunk 2]
```diff
@@ -0,0 +1,5 @@
+- `[mempool]` Application can now set `ConsensusParams.Block.MaxBytes` to -1
+  to gain more control on the max size of transactions in a block.
+  It also allows the application to have visibility on all transactions in the
```
[the remark this answers is not in the document]
Comment (sergio-mena) [reply .changelog/unreleased/improvements/980-max-size-more-control.md:3]: Nope. At `PrepareProposal` time, we don't put any limit when we reap the transactions. That's why I added the warning in the "breaking changes" part: app might be _forced_ to remove some of the TXs cometBFT is sending up when filling `ReponsePrepareProposal`.

File: spec/abci/abci++_app_requirements.md [hunk 3]
```diff
@@ -600,7 +601,19 @@ This is enforced by the consensus algorithm.
...
+of transactions it returns, and by `ProcessProposal` to reject any received block
+whose total transaction size is bigger than the enforced limit.
+In such case, the Application MAY set `MaxBytes` to -1.
+
+If the Application sets value -1, consensus will:
+
+- consider that the actual value to enforce is 100 MB
+- will provide *all* transactions in the mempool in calls to `PrepareProposal`
+
+Must not be 0; must have `-1 < MaxBytes < 100 MB`.
```
[the remark this answers is not in the document]
Comment (sergio-mena) [reply spec/abci/abci++_app_requirements.md:616]: Good catch: '<='

File: spec/consensus/creating-proposal.md [hunk 4]
```diff
@@ -4,40 +4,58 @@ order: 2
...
-amino overhead for a block (`MaxAminoOverheadForBlock`), the size of
+protobuf overhead for a block (`MaxOverheadForBlock`), the size of
 the last commit (if present) and evidence (if present). While reaping
-we account for amino overhead for each transaction.
+we account for protobuf overhead for each transaction.
 
 ```go
-func MaxDataBytes(maxBytes int64, valsCount, evidenceCount int) int64 {
- return maxBytes -
+func MaxDataBytes(maxBytes, evidenceBytes int64, valsCountint) int64 {
```
[the remark this answers is not in the document]
Comment (sergio-mena) [reply spec/consensus/creating-proposal.md:22]: Oops!

File: spec/consensus/creating-proposal.md [hunk 5]
```diff
@@ -4,40 +4,58 @@ order: 2
...
 
+If `ConsensusParams.Block.MaxBytes == -1`, we reap *all* outstanding transactions from the mempool
+
+## Preparing the proposal
+
+Once the transactions have been reaped from the mempool according to the rules described above,
+CometBFT calls `PrepareProposal` to the application with the transaction list that has just been reaped.
+As part of this call the application can remove, add, or reorder transactions in the transaction list.
+
+The `RequestPrepareProposal` contains to important fields:
```
[the remark this answers is not in the document]
Comment (sergio-mena) [reply spec/consensus/creating-proposal.md:39]: Oops again! :-)

File: types/params_test.go [hunk 6]
```diff
@@ -39,6 +39,7 @@ func TestConsensusParamsValidation(t *testing.T) {
 		11: {makeParams(1, 0, 2, 0, []string{}, 0), false},
 		// test invalid pubkey type provided
 		12: {makeParams(1, 0, 2, 0, []string{"potatoes make good pubkeys"}, 0), false},
+		13: {makeParams(-1, 0, 2, 0, valEd25519, 0), true},
```
[the remark this answers is not in the document]
Comment (sergio-mena) [reply types/params_test.go:42]: Thanks for this!

**Pull request merged**